Uwd8kvpl trigger the metadata publish event on state changes

jkunle opened the pull request:
Continuation of [modelling signing certificates card](https://trello.com/c/ee2PV5rg/348-model-signing-certificate-rotation-in-the-app) and follow up to [PR add events for enabling and disabling signing certificates](https://trello.com/c/iCTTeo3L/373-add-events-for-enabling-and-disabling-signing-certificates)

WHAT
All the new events that change the state of certificates should trigger the metadata publish event.

Replace certificate#update with separate enable and disable actions.

WHY
Refer to links supplied

chrisholmes commented on app/controllers/certificates_controller.rb:
```diff
@@ -14,14 +14,15 @@ def create
-  def update
+  def enable
     @certificate = Certificate.find(params[:id])
+    EnableSigningCertificateEvent.create(certificate: @certificate)
```
We should probably add some error handling here.

What should have when the state change fails and #create returns false ?

chrisholmes commented on app/controllers/certificates_controller.rb:
```diff
@@ -37,6 +38,6 @@ def upload_params
 
   def update_params
```
I think we can remove this method now right?

chrisholmes commented on app/models/signing_certificate_event.rb:
```diff
@@ -4,6 +4,7 @@ class SigningCertificateEvent < AggregatedEvent
   belongs_to_aggregate :certificate
   validates_presence_of :certificate
   validate :certificate_is_signing?
+  after_save :trigger_publish_event
```
I think we should consider introducing a class_method for this so keep the code DRY.

chrisholmes commented on app/views/components/show.html.erb:
```diff
@@ -6,11 +6,11 @@
+<%= render "certificates/certificates_table" , status: 'disable' %>
 
 <h3 class="govuk-heading-m">Signing Certificates (Disabled)</h3>
 <% @certificates =  @component.disabled_signing_certificates %>
```
reassigning a ivar inside a view is bit smelly. Could we pass this as a parameter to the partial like `status`?
```
 <%= render "certificates/certificates_table" , 
     status: 'enable', 
     certificates: @component.enabled_signing_certificates
 %>
```

chrisholmes commented on app/views/certificates/_certificates_table.html.erb:
```diff
@@ -16,9 +16,10 @@
-          <%= f.submit certificate.enabled ? 'Disable' : 'Enable', class: 'govuk-button' %>
+        <%= form_for [@component, certificate],
+          method: :patch,
+          url: "/components/#{@component.id}/certificates/#{certificate.id}/#{status}"  do |f| %>
```
you should be able to use a url_helper for this

jkunle replied:
It can but not sure if this is really worth it. Had to do the following
```
  def toggle_enable_on_button(status, component, certificate)
    helper = Rails.application.routes.url_helpers
    name = "#{status}_component_certificate_path"
    helper.public_send(name, component, certificate) if helper.respond_to?(name)
  end
```
to get this to work

jkunle commented:
Made some suggested changes

- **Changed SigningCertificateEvent** to **ToggleEnableEvent**
- **DRY** out **PublishServicesMetadataEvent.Create** method calls into Class_Method **TriggerMetadataEventCallback.publish**
- Removed certificate metadata from data: {} area of Event produced from **enable** and **disable** events
- Introduced feedback when enable/disable button are clicked but nothing actually happens
**### Note:** 
Made change to metadata produced on publish but only picking the **enabled_signing_certificates** as opposed to all **signing_certificates**

chrisholmes commented on app/controllers/certificates_controller.rb:
```diff
@@ -14,13 +14,24 @@ def create
+    event = EnableSigningCertificateEvent.create(certificate: @certificate)
+    unless event.valid?
+      error_message = event.errors.full_messages
+      Rails.logger.error(error_message)
```
We probably don't need to log here, but we can look at logging later.

chrisholmes commented on app/models/toggle_Enable_event.rb:
```diff
@@ -0,0 +1,20 @@
+  private
+
+  def certificate_is_signing?
+    return unless certificate.usage == 'encryption'
```
I think should probably be a check like:

```suggestion
    return if certificate.usage == 'signing'
```

chrisholmes replied:
Alternatively, we could add some helpers to the `Certifcate` class like '#signing?' so that we can abstract away the data model.

jakubmiarka commented on app/models/toggle_Enable_event.rb:
```diff
@@ -0,0 +1,20 @@
+class ToggleEnableEvent < AggregatedEvent
```
could we fix casing of the filename?
